app/feature/feature-chat/src/main/kotlin/com/hedvig/android/feature/chat/ChatViewModelNew.kt

```diff
@@ -0,0 +1,161 @@
+  ChatUiState(),
+  ChatPresenter(
+    chatRepository = chatRepository,
+    chatClosedTracker = chatClosedTracker,
+    featureManager = featureManager,
+  ),
+) {
+  init {
+    emit(ChatEventNew.FetchMessages())
+    viewModelScope.launch {
+      while (true) {
+        delay(5.seconds)
+        emit(ChatEventNew.PollMessages)
+      }
+    }
+  }
+}
+
+internal class ChatPresenter(
+  private val chatRepository: Provider<ChatRepositoryNew>,
+  private val chatClosedTracker: ChatEventStore,
+  private val featureManager: FeatureManager,
+) : MoleculePresenter<ChatEventNew, ChatUiState> {
+
+  @Composable
+  override fun MoleculePresenterScope<ChatEventNew>.present(lastState: ChatUiState): ChatUiState {
+    var isLoadingChat by remember { mutableStateOf(lastState.isLoadingChat) }
+    var isLoadingNewMessages by remember { mutableStateOf(lastState.isLoadingNewMessages) }
+    var messages by remember { mutableStateOf(lastState.messages) }
+    var isSendingMessage by remember { mutableStateOf(lastState.isSendingMessage) }
+    var canFetchMoreMessages by remember { mutableStateOf(lastState.canFetchMoreMessages) }
+    var canFetchUntil by remember { mutableStateOf(lastState.canFetchUntil) }
+    var errorMessage by remember { mutableStateOf(lastState.errorMessage) }
+
+    var fetchId by remember { mutableIntStateOf(0) }
+    var fetchUntil by remember { mutableStateOf<Instant?>(null) }
+    var fileState by remember { mutableStateOf<File?>(null) }
+    var messageInputState by remember { mutableStateOf<String?>(null) }
+
+    CollectEvents { event ->
+      when (event) {
+        is ChatEventNew.FetchMessages -> fetchUntil = event.until
+        ChatEventNew.PollMessages -> fetchId++
+        is ChatEventNew.SendFileMessage -> fileState = event.file
+        is ChatEventNew.SendTextMessage -> messageInputState = event.message
+        is ChatEventNew.DismissError -> errorMessage = null
+      }
+    }
+
+    LaunchedEffect(fetchId, fetchUntil) {
+      isLoadingNewMessages = fetchUntil != null
+      chatRepository.provide()
+        .fetchChatMessages(fetchUntil)
+        .fold(
+          ifLeft = {
+            errorMessage = it.message
+            isLoadingChat = false
+            isLoadingNewMessages = false
+          },
+          ifRight = {
+            messages = it.messages.toPersistentList()
+            canFetchMoreMessages = it.hasNext
+            canFetchUntil = it.nextUntil
+            isLoadingChat = false
+            isLoadingNewMessages = false
+          },
+        )
+    }
+
+    LaunchedEffect(fileState) {
+      val file = fileState ?: return@LaunchedEffect
+      isSendingMessage = true
+      chatRepository.provide()
+        .sendFile(file, "")
+        .fold(
+          ifLeft = {
+            errorMessage = it.message
+            isSendingMessage = false
+          },
+          ifRight = {
+            it.message?.let(messages::add)
+            isSendingMessage = false
+          },
+        )
```

Comment:
Should we be clearing the `fileState` after this effect is run? So we know it's now handled and is null again.
Same for `messageInputState`